### ISSUE TITLE
[feature request] `esc` key binding to de-select (and avoid an extra click :) ?

### ISSUE THREAD

**Issue by fdch** [pure-data/pure-data#547]
It would be nice to have something like `esc` key to de-select. Right now, doing `Select All` twice does toggle between selection and no selection. However, when there is a range of objects selected you have to either click on the canvas or toggle edit mode.

The reason I ask for this is that given the new feature for intelligent patching, I often have a group of selected objects, and I need to deselect them before making any connections, otherwise I end up with a bunch of unwanted connections.  But it would be nice to have just one way to do that with the keyboard-hand on the fly.

**Comment by umlaeute** [comment]
the solution i have in mind for your underlying problem (intelligent patching getting in the way) is more like: having a modifier key to *bypass* intelligent patching.

as in:
- select multiple objects
- connect one of them to an unselected object -> all connect
- press <kbd>Shift</kbd> and keep it pressed
- connect one of the selected objects to an unselected object -> only the directly asked for connection is being made.

i'm mostly unsure about the correct key to temporarily disable intelligent patching.

and i wonder whether (de)selections should be undoable as well (but that would be another feature request).

**Comment by fdch** [comment]
Imho, I would not suggest to bypass the feature, which is altogether quite amazing. 

For example, what other situation would one need to bypass it?

I probably have not used it enough to think of any. 

However, I do find often the need to deselect stuff, and intelligent patching was an example (and my own pd_user_baggage was getting in the way ;)

**Comment by umlaeute** [comment]
i already have this feature request for being able to *temporarily* disable intelligent patching (in private @tmusil !!!).

i just don't see that much difference between "clicking on the canvas" and "hitting <kbd>ESC</kbd>"

**Comment by fdch** [comment]
The difference as I see it is that, after selecting:

in one combined action I can:
-  move the pointer straight to the outlet while hitting `esc` (only once)
-  then just click the already deselected outlet.

Otherwise, I have to
-  click on the canvas
-  move the pointer to the outlet
-  click the deselected outlet

**Comment by umlaeute** [comment]
you're cheating a bit by putting two actions into one line in your first example.
as i see it:
- hit <kbd>Esc</kbd> **or** click on the canvas
- move the pointer straight to the outlet
- click the outlet

there's not *so much* difference here.

the option i'm suggesting is:
- move the pointer straight to the outlet
- hit <kbd>Shift</kbd>, at the same time do:
- click the outlet

again, we have the same number of steps with comparable complexity.
the key difference is that the objects are still selected (so you can do other things with them, like keep moving them somewhere else)

**Comment by fdch** [comment]
You are indeed right, and I had not considered that bonus of having the
selection still in place!

I think the esc key would then not be suitable for this situation... and
your solution allows for more possibilities indeed.

Thanks for looking into this :)

**Comment by musger** [comment]
I would like to suggest to introduce a modifier key to enable the connect anything-to-anything behavior (which rarely does semantically correct connections imho) instead of introducing modifiers to temporarily disable this specialist behavior or additional mouse/key clicks to deselect objects as a workaround?

**Comment by ch-nry** [comment]
I'm jumping a bit late in this thread. I just want to rise few points : 
- shit is used for object creation in french keyboard since number are not directlly accessible. So french used need to "control shift 1" to create an object. This must be simplified if shift should be used as a modifier to enable or disable a feature.  A 4 key shortcut seems a bit complex.
- A keyboard shortcut is a lot's faster action than a mouse click because pd user hands are already a lot's on the keyboard. Having the hand to jump from the keyboard to the mouse is not really efficient.
- having a 100% keyboard solution to patch would be a huge benefit for the workflow.

**Comment by umlaeute** [comment]
thanks.
i think if we are only talking about connections ("intelligent patching" is *mostly* about connecting multiple objects/iolets in a single run), using <kbd>Shift</kbd> should be ok, even on french keyboards.

the only "intelligent patching" features that are related to object-creation are
- autopatching
- ... (anything else?)
